### ISSUE TITLE
Arm64 build Raspberry Pi

### ISSUE THREAD

**Issue by CarloCattano** [acids-ircam/nn_tilde#3]
Having trouble getting it to compile on a pi 3 ``` 5.15.30-v8+aarch64 Raspberry Pi OS```

Just modified the CMakeLists and removed a compiler flag that wasnt recognized ( same goes for an Ubuntu PC ) did not know what was a good replacement flag for it . On ubuntu I could compile succesfully tho .

got the libtorch files from nigthly and tried also with github.com/ljk53/pytorch-rpi provided libs for arm , and no difference

```
/usr/bin/cmake -S/home/pi/nn_tilde/src -B/home/pi/nn_tilde/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /home/pi/nn_tilde/build/CMakeFiles /home/pi/nn_tilde/build//CMakeFiles/progress.marks
make  -f CMakeFiles/Makefile2 all
make[1]: Entering directory '/home/pi/nn_tilde/build'
make  -f backend/CMakeFiles/backend.dir/build.make backend/CMakeFiles/backend.dir/depend
make[2]: Entering directory '/home/pi/nn_tilde/build'
cd /home/pi/nn_tilde/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/pi/nn_tilde/src /home/pi/nn_tilde/src/backend /home/pi/nn_tilde/build /home/pi/nn_tilde/build/backend /home/pi/nn_tilde/build/backend/CMakeFiles/backend.dir/DependInfo.cmake --color=
make[2]: Leaving directory '/home/pi/nn_tilde/build'
make  -f backend/CMakeFiles/backend.dir/build.make backend/CMakeFiles/backend.dir/build
make[2]: Entering directory '/home/pi/nn_tilde/build'
[ 25%] Building CXX object backend/CMakeFiles/backend.dir/backend.cpp.o
cd /home/pi/nn_tilde/build/backend && /usr/bin/c++ -DUSE_C10D_GLOO -DUSE_DISTRIBUTED -DUSE_RPC -DUSE_TENSORPIPE -isystem /home/pi/libtorch/include -isystem /home/pi/libtorch/include/torch/csrc/api/include -D_GLIBCXX_USE_CXX11_ABI=0 -O3 -DNDEBUG -fPIC -D_GLIBCXX_USE_CXX11_ABI=0 -std=gnu++14 -o CMakeFiles/backend.dir/backend.cpp.o -c /home/pi/nn_tilde/src/backend/backend.cpp
``` 
It gets stuck on 25% forever.

Would be nice to test this on the rpi3 and 4 with the pisound hat and see how good it performs.

Thanks for any hint

**Comment by CarloCattano** [comment]
``` '-faligned-allocation'; did you mean '-fsized-deallocation'? ```
this was the error that lead me to remove the flag

**Comment by CarloCattano** [comment]
on a Pi 4 8 Gb ram model Im able to build ( using https://github.com/Qengineering/PyTorch-Raspberry-Pi-64-OS  cpu version ) 

But when making a pd patch the object doesnt create . Only warning in pd console is "..couldn't create"
After setting verbose I understood its missing libmpi_cxx.so , libopenblas.so.

After this the object loads as it should but it raises errors on the verbose output and the inlets outlets dont change for encode/decode

[debugOut.txt](https://github.com/acids-ircam/nn_tilde/files/8454523/debugOut.txt)

**Comment by caillonantoine** [comment]
Make sure that you have a [64bit raspios](https://downloads.raspberrypi.org/raspios_arm64/images/)
Install [this version of miniforge](https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge3-Linux-aarch64.sh) on your Pi
Install pytorch `pip install torch`
Locate libtorch (usually something like this `/path/to/miniforge3/lib/python3.9/site-packages/torch`), and then build the project
Using this method the external is working well on my RPi, tell me if you still have problems !

**Comment by caillonantoine** [comment]
Have you tried loading your pre-trained model somewhere else ? (nn~ for macos or linux ? )

**Comment by CarloCattano** [comment]
Hey thanks for the answer!
I'm on latest RaspiOS 64bit bullseye. 
Yes the model loads fine on Ubuntu PC and have tried also with the wheel dataset. 
Will do as suggested and will probably run fine, can't wait to test this with my PiSound hat and see how well it goes🤙
Will try in a week when back home

**Comment by CarloCattano** [comment]
> Locate libtorch (usually something like this `/path/to/miniforge3/lib/python3.9/site-packages/torch`), and

weird , after installing miniforge , that path site-packages doesn't contain torch...
when running ```pip show torch ``` the route is on my normal python 3.9 installation directory .

If I compile the object with that path Im still unable to run it on pd ( no inlets , outlets and a crash when dsp on with similar verbose output as before) 

[cmakeWarnings.txt](https://github.com/acids-ircam/nn_tilde/files/8584953/cmakeWarnings.txt)
[miniforge.txt](https://github.com/acids-ircam/nn_tilde/files/8585022/miniforge.txt)

**Comment by CarloCattano** [comment]
@caillonantoine  My bad !  a tiny typo on the path of the model  
However still not using the torch from the miniforge path but the usual from pip install torch regularly 
Can consider the issue closed since it works and will jump into #5   to learn how to optimize it to run .

Even with buffers like 32768 I get the 4 cores of the pi at 100% and must kill it manually and temperatures rising up quickly too

Want to investigate if it has to do with finding an replacement for '-faligned-allocation' flag for aarch64 and see if that helps performance
